### ISSUE TITLE
False positive, when throw happens in closure

### ISSUE THREAD

**Issue by aik099** [squizlabs/PHP_CodeSniffer#738]
I'm getting `Missing @throws tag in function comment` error for outer function in following code:

```php
/**
 * Sets expectation for specific command.
 *
 * @param boolean $is_interactive Is interactive.
 *
 * @return void
 */
private function _expectCommand($is_interactive = false)
{
	$expectation->will(function () use ($process) {
		throw new ProcessFailedException($process->reveal());
	});
}
```

The sniff that reports the error should ignore `throw` statements without a closures.

**Comment by gsherwood** [comment]
Just because a throw statement is inside a closure doesn't mean it wont be thrown when you call this function. In your specific case, maybe the closure code wont actually be executed when you call _expectCommand, but there is no way for PHPCS to know this.

So if I ignore all throw statements inside closures, I'm potentially going to not be reporting errors in cases where I should be.

Does that make sense?

**Comment by aik099** [comment]
I'm afraid it does. In similar case in past where exception was catched inside the method and rethrown in IF statement it was also better to consider exception as always thrown in that method.

Since we don't have any way of detecting if closure if called right in here, then it's better to keep everything as-is.

**Comment by gsherwood** [comment]
> Since we don't have any way of detecting if closure if called right in here, then it's better to keep everything as-is.

Agreed. It's not great, but I can't think of a way around it.

**Comment by mindplay-dk** [comment]
In my opinion, you've got this backwards.

The exception is thrown inside a closure - in other words, it isn't being thrown by the parent method, but by the closure.

If anything, I should have to annotate the closure, not the method, with `@throws`.

This is no different from a situation like:

```php
class Foo
{
    public function hello() {
        $this->boom();
    }

    protected function boom() {
        throw new RuntimeException();
    }
}
```

Should I have to annotate `boom()` with `@throws` in this case? It will, invariably, cause an exception - although there is no direct way to know that.

The failed inspection should alert me to situations where we *do* know an exception will be thrown.

In the case where a closure gets defined inside a method, all we know about the method, is that it defines a closure. That's it. We don't know whether it does or doesn't thrown an exception.

Case in point:

```php
class Foo {
    /**
     * @throws RuntimeException ... well, no, it doesn't, this tag just makes CS happy
     */
    public function getFunc() {
        return function() { throw new RuntimeException(); }
    }
}
```

We actually have comments on `@throws` tags in our codebase with descriptions like this one.

Being forced to add an incorrect `@throws` tag is extremely misleading and makes the consumer of that method have to stop and read documentation explaining how it *doesn't* throw.

You're betting on the fact that any closure being defined, which throws an exception, is also going to be called within the same scope - which, if anything, is more likely to be wrong than right. Rarely does anyone define closures just to call them within the same scope - they're usually being passed somewhere else, where they might cause an exception, and you can't catch that case either.

Consumers of closures can sometimes trigger an exception - that's the reality of using closures.

Fortunately, in the large majority of cases, exceptions thrown from within closures are RuntimeExceptions - they're not exceptions intended to be handled, they being thrown to report unrecoverable errors, which maybe only an application-wide error-handler or error-reporting facility would trap and recover from by logging or outputting an error-message.

Just to demonstrate the worst-case example:

```php
class Foo
{
    public function hello() {
        $boomer = $this->getBoomer();
        $boomer();
    }

    public function getBoomer() {
        return function() { throw new RuntimeException(); };
    }
}
```

In this case `hello()` results in an exception, for which there is no inspection - and `getBoomer()`, which *can't* throw an exception, gets a failed inspection.

Two out of two wrong guesses, which demonstrates the problem: you're just guessing.

This example isn't fictive, by the way - I've run into precisely this situation more than once.

**Comment by mindplay-dk** [comment]
At the very least, there should be an option to toggle the scope of this inspection, which, as demonstrated, does produce false positives.

The current hehavior could of course be preserved as an option, but I don't think a setting that delivers false positives should be the default?
